Quick Start for Existing Users V2: Assign new tour only to new existing users

ashiagr opened the pull request:
Parent #16350

This PR adds common tasks: `check_stats`, `view_site`, `follow_site` to a new `Quick Start` card and displays it only to new existing users.

Corresponding FluxC PR: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2387

https://user-images.githubusercontent.com/1405144/168042785-5c87a899-2866-43f6-bdfe-eb2d21cfed56.mov


To test: 

/cc @jostnes, @osullivanchris 

#### Test.1 Login Epilogue - `QuickStartForExistingUsersV2FeatureConfig` disabled - Displays old `Quick Start` card

1. Launch the app.
2. Logout/login and select "Show me around" when the quick start prompt is shown.
3. Select an existing site.
4. ✅ Notice that the old Quick Start card with `Customize` and `Grow` task types is shown.
5. Logout/login and select "Show me around" when the quick start prompt is shown.
6. Create a new site.
7. ✅ Notice that the old Quick Start card with `Customize` and `Grow` task types is shown.

#### Test.2 Site Picker - `QuickStartForExistingUsersV2FeatureConfig` disabled - Displays old `Quick Start` card

1. Continue from Test.1
2. Create a new site using the site picker
3. ✅ Notice that the old Quick Start card with `Customize` and `Grow` task types is shown

#### Test.3 Login Epilogue - `QuickStartForExistingUsersV2FeatureConfig` enabled - Displays new `Quick Start` card when site is selected and old `Quick Start` card when site is created

1. Launch the app
2. Enable to `Me` -> `App Settings` -> `Debug settings` -> `QuickStartForExistingUsersV2FeatureConfig`
3. Restart the app.
4. Logout/login and select "Show me around" when the quick start prompt is shown.
5. Select an existing site
6. ✅ Notice that the new `Quick Start` card with `Get to know the WordPress app`  task type is shown. 
7. Logout/login and select "Show me around" when the quick start prompt is shown.
8. Create a new site.
9. ✅ Notice that the old `Quick Start` card with `Customize` and `Grow` task types is shown.

#### Test.4 Site Picker - `QuickStartForExistingUsersV2FeatureConfig` enabled - Displays old `Quick Start` card when site is created.

1. Continue from Test.3.
2. Create a new site using the site picker.
3. ✅ Notice that the old Quick Start card with `Customize` and `Grow` task types is shown

#### Test.5 New Quick Start card - Common Tasks

1. ✅ Notice that the new Quick Start card with “Get to know the WordPress app”  task type currently has only three tasks that are common to the old Quick Start card: `check_stats`, `view_site`, `follow_site`.
2. Run tests in [wpandroid-quick-start-testing-instructions.txt](https://github.com/wordpress-mobile/WordPress-Android/files/8246096/wpandroid-quick-start-testing-instructions.txt).
3. ✅ Notice that all tests pass.

## Merge Instructions

1. Make sure `FluxC` PR is merged to trunk ✅ 
2. Update `FluxC` version in build.gradle ✅ 
3. Remove `Not Ready for Merge` label ✅ 
4. Merge the PR

## Regression Notes
1. Potential unintended areas of impact
Wrong `Quick Start Card` is shown based on whether site selected or created from login epilogue.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
See `To test` section

3. What automated tests I added (or what prevented me from doing so)
- Added unit tests for start quick start based on existing/ new site. - e6e9563
- Added unit tests for `check_stats`, `view_site` tasks for the new Quick Start card. - 6651903
- Yet to add tests for `follow_site` task.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Add labels to QuickStartType

ashiagr committed:
Add deletable pref key for last selected quick start type with getter/ setter

ashiagr committed:
Set key in intent to identify that site is created from login epilogue

ashiagr committed:
Set key in intent to identify that landing editor is opened for new site

ashiagr committed:
Get and set last selected quick start type from prefs in qs repo

ashiagr committed:
Pass new site info from fragment to qs repository to set correct qs type

ashiagr committed:
Tests: Fix existing tests

ashiagr committed:
Tests: Add matching tests to test checkAndStartQuickStart(...) for existing site

ashiagr committed:
Merge branch 'trunk' into issue/16350-add-tasks-for-existing-site-task-type-wpandroid

ashiagr committed:
Handle check_stats and view_sites for existing site qs type

ashiagr committed:
Tests: Add tests for existing site QS view_site and check_stats tasks

ashiagr committed:
Handle follow_site task for existing site qs type

ashiagr committed:
Fix test after merge

ashiagr committed:
Point to FluxC PR

ashiagr committed:
Merge branch 'trunk' into issue/16350-add-tasks-for-existing-site-task-type-wpandroid

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/quickstart/QuickStartRepository.kt
#	WordPress/src/test/java/org/wordpress/android/ui/mysite/cards/quickstart/QuickStartRepositoryTest.kt

ashiagr committed:
Fix wp main activity existing tests

ashiagr committed:
Ktlint: Fix spacing

ashiagr committed:
Dismiss reader qs notification on changing tabs

ashiagr committed:
Update reader prompt dismiss logic

ashiagr committed:
Merge branch 'trunk' into issue/16350-assign-new-tour-only-to-new-existing-users-wpandroid

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/quickstart/QuickStartRepository.kt
#	WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt
#	build.gradle

ashiagr committed:
Fix merge conflict

ashiagr committed:
Merge branch 'trunk' into issue/16350-assign-new-tour-only-to-new-existing-users-wpandroid

ashiagr committed:
Point to FluxC PR with merged trunk

ashiagr committed:
Add existing site check stats task to tab menu step

ashiagr committed:
Merge branch 'trunk' into issue/16350-assign-new-tour-only-to-new-existing-users-wpandroid

# Conflicts:
#	WordPress/src/test/java/org/wordpress/android/ui/mysite/items/SiteItemsBuilderTest.kt

ashiagr committed:
Update existing site check stats task to tab menu step

jostnes commented:
Did some testing on this build and overall the feature flag looks to be working as expected as per the above scenarios. I have two observations to share from testing:
1. My test phone is Oppo Reno 6Z (Google says that the screen size is 6.43-inch) but on this phone, the text on the new card seems to be too long, overlapping with the "Done" icon:
![qsv2](https://user-images.githubusercontent.com/17252150/168224476-bc4408e8-2117-4ff6-821b-dfd13a3ed13b.jpg)

2. When I restarted the app after updating the feature flag, I still get the previous QS version. It will only be updated to the correct one after I logout/relogin. I'm imagining the case where, when this feature is turned on for users, users who are already seeing the older QS should continue seeing it, is that correct? And only users who go through the Show me Around prompt will get to see the new version?

AjeshRPai commented on WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartType.kt:
```diff
@@ -50,4 +50,17 @@ sealed class QuickStartType(val tasks: List<QuickStartTask>) {
+        private const val NEW_SITE = "new-site"
+        private const val EXISTING_SITE = "existing-site"
+
+        fun fromString(label: String): QuickStartType {
```
Nitpick:  I think we can rename the function to `fromLabel`, WDYT? 🤔

ashiagr replied:
Renamed in a97be94

AjeshRPai reviewed (approved):
@ashiagr I have tested the code changes according to the test instructions. All the changes are good. 👍🏼 🙌🏼 

I have left a nitpick for you to consider, as that is non blocking I approving the PR. Please feel free to merge once you have addressed it.

ashiagr committed:
Update fluxc version

ashiagr committed:
Rename method

ashiagr commented:
Thanks, @AjeshRPai 

I've addressed your comment in a97be94.

Thanks, @jostnes. Please see my responses below:

> 1. My test phone is Oppo Reno 6Z (Google says that the screen size is 6.43-inch) but on this phone, the text on the new card seems to be too long, overlapping with the "Done" icon:

I've created an issue https://github.com/wordpress-mobile/WordPress-Android/issues/16539 to track it.

> 2. When I restarted the app after updating the feature flag, I still get the previous QS version. It will only be updated to the correct one after I logout/relogin. I'm imagining the case where, when this feature is turned on for users, users who are already seeing the older QS should continue seeing it, is that correct? And only users who go through the Show me Around prompt will get to see the new version?

This is correct. For more details, you can refer to this internal discussion: p1649849406838359-slack-C027K4MNPGQ

Since all comments are non-blocking, I'll go ahead and merge the PR once CI completes.

osullivanchris commented:
Took a look at the build and all working as expected so far 👍 😄
